ostd/src/mm/kspace/kva.rs

```diff
@@ -76,138 +88,341 @@ impl VirtAddrAllocator {
+    pub unsafe fn map_untracked_pages(
+        &mut self,
+        va_range: Range<Vaddr>,
+        pa_range: Range<Paddr>,
+        prop: PageProperty,
+    ) {
+        assert!(pa_range.start % PAGE_SIZE == 0);
+        assert!(pa_range.end % PAGE_SIZE == 0);
+        assert!(va_range.len() == pa_range.len());
+        assert!(self.start() <= va_range.start && self.end() >= va_range.end);
 
-    /// Get the mapped page.
-    /// This function will fail in the following cases:
-    /// * if the provided page type doesn't match the actual mapped one.
-    /// * If the address is not mapped (`NotMapped`), the function will fail.
-    /// * If the address is mapped to a `MappedUntracked` page, the function will fail.
-    pub fn get_page<T: PageMeta>(&self, addr: Vaddr) -> Result<Page<T>> {
-        assert!(self.start() <= addr && self.end() >= addr);
-        let start = addr.align_down(PAGE_SIZE);
-        let vaddr = start..start + PAGE_SIZE;
         let page_table = KERNEL_PAGE_TABLE.get().unwrap();
-        let mut cursor = page_table.cursor(&vaddr).unwrap();
-        let query_result = cursor.query().unwrap();
+        let mut cursor = page_table.cursor_mut(&va_range).unwrap();
+        let flusher = TlbFlusher::new(CpuSet::new_full(), disable_preempt());
+        // SAFETY: The caller of `map_untracked_pages` has ensured the safety of this mapping.
+        unsafe {
+            cursor.map_pa(&pa_range, prop);
+        }
+        flusher.issue_tlb_flush(TlbFlushOp::Range(va_range.clone()));
+        flusher.dispatch_tlb_flush();
+    }
+
+    /// Gets the mapped untracked page.
+    ///
+    /// This function returns None if the address is not mapped (`NotMapped`),
+    /// while panics if the address is mapped to a `Mapped` or `PageTableNode` page.
+    pub fn get_untracked_page(&self, addr: Vaddr) -> Option<(Paddr, usize)> {
+        let query_result = self.query_page(addr);
         match query_result {
-            PageTableItem::Mapped {
+            PageTableItem::MappedUntracked {
                 va: _,
-                page,
+                pa,
+                len,
                 prop: _,
-            } => {
-                let result = Page::<T>::try_from(page);
-                if let Ok(page) = result {
-                    Ok(page)
-                } else {
-                    panic!("the provided page type doesn't match the actual mapped one");
-                }
-            }
+            } => Some((pa, len)),
+            PageTableItem::NotMapped { .. } => None,
             _ => {
                 panic!(
-                    "Unexpected query result: Expected 'Mapped', found '{:?}'",
+                    "Found '{:?}' mapped into untracked `Kva`, expected `MappedUntracked`",
                     query_result
                 );
             }
         }
     }
 }
 
-impl Drop for Kva {
+impl<M: AllocatorSelector + 'static> Drop for Kva<M> {
     fn drop(&mut self) {
```

Comment:
This `drop` method does quite some work. I am wondering if it has negative effect on the performance of creating/destroying kernel threads, and thus affecting the LMbench. Could you show the results of LMbench and see if there are any regression?

Reply:
The results of sysbench/thread_lat:

Average Execution Time per Thread on Linux:
- linux: 16.92 ms
- asterinas main: 25.58 ms
- asterinas kva: 26.47 ms

The results of lmbench/process_fork_lat:

Average Fork latency on Linux:
- linux: 95.1034 µs
- asterinas main: 87.4516 µs
- asterinas kva: 95.8571 µs

Considering that using `Kva` to manage the kernel stack can achieve better page protection mechanisms and improve safety, I think the small regression is acceptable. After merging this PR, we will make the algorithm for kernel virtual memory allocation parallel and optimize it to $O(\log n)$, which will also improve the efficiency of creating/destroying kernel threads.

Reply:
It looks the overheads are acceptable.

Reply:
Forgot to check Makefile. In the previous results, I forgot to remove the following lines in the `Makefile`:

https://github.com/asterinas/asterinas/blob/2af9916de92f8ca1e694bb6ac5e33111bbcf51fd/Makefile#L71-L72

After removing the lines, `OSTD_TASK_STACK_SIZE_IN_PAGES` should be 8. The new results are as follows:

The results of sysbench/thread_lat:

Average Execution Time per Thread on Linux:

 -linux: 16.84 ms
- asterinas main: 26.69 ms
- asterinas kva: 26.60 ms

The results of lmbench/process_fork_lat:

Average Fork latency on Linux:

- linux: 95.8966 µs
- asterinas main: 99.9057 µs
- asterinas kva: 96.5893 µs

There is no negative effect on the performance of the above benchmarks for enabling kernel virtual area.